File: src/parse_requests.py
# Growth dataset -> fixed service duration and area of service while the number of requests varies from 50 to 1000
from typing import Union
"""
SrcPoint_x= The x-coordinate of customer pick-up point.

SrcPoint_y= The y-coordinate of customer pick-up point.

DstPoint_x= The x-coordinate of customer drop-off point.

DstPoint_y= The y-coordinate of customer drop-off point.

DRT= Direct Ride Time. The time needed to transport the customer between Src and Dst point on the shortest path without any deviations.

MRT= Maximum Ride Time (calculated by adding 60 minutes to the direct ride time of the customer).

Pick-up/Drop-up time= Pickup or drop-off time of the customer determined by the flag.

Flag= A Boolean value that indicates if the time in previous column is pick-up (1) or drop-off (0) time.
"""

ALPHA = 0.8  # Alpha coefficient, used to calculate the dropoff time window
IDX = 0
SRC_POINT_X = 1
SRC_POINT_Y = 2
DST_POINT_X = 3
DST_POINT_Y = 4
DRT = 5
MRT= 6
PICKUP_OR_DROPOFF_TIME = 7
ISPICK_TIME = 8

class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

class Request:
    def __init__(self,id, line: str, service_duration: int, area_of_service: int):

        self.service_duration = service_duration
        self.area_of_service = area_of_service

        parts = filter(lambda x: x!="", line.split("\t"))
        for i, x in enumerate(parts):
            x = x.strip("\t")
            val = int(x)
            if i == IDX:
                self.id = id
            elif i == SRC_POINT_X:
                self.src_x = val
            elif i == SRC_POINT_Y:
                self.src_y = val
            elif i == DST_POINT_X:
                self.dst_x = val
            elif i == DST_POINT_Y:
                self.dst_y = val
            elif i == DRT:
                self.drt = val
            elif i == MRT:
                self.mrt = val
            elif i == PICKUP_OR_DROPOFF_TIME:
                self.__time = val
            elif i == ISPICK_TIME:
                self.pickup_time = self.__time
                # dropoff = alpha * maxride + pickup
                self.dropoff_time = int(ALPHA * self.mrt) + self.pickup_time

    def src_point(self):
        return (self.src_x, self.src_y)

    def dst_point(self):
        return (self.dst_x, self.dst_y)

    def points(self):
        return [self.src_point(), self.dst_point()]

    def __str__(self):
        src = f"SRC=({self.src_x}, {self.src_y})"
        dst = f"DST=({self.dst_x}, {self.dst_y})"
        drt = f"DRT={self.drt}"
        mrt = f"MRT={self.mrt}"
        picktime = f"PICKTIME={self.pickup_time}"
        sd = f"SERVICE_DURATION={self.service_duration} min"
        aos = f"AREA_OF_SERVICE={self.area_of_service} miles"

        return " | ".join([src, dst, drt, mrt, picktime, sd, aos])

def sqKm_to_sqMiles(km: float) -> float:
    return km * 0.386102

def getRequests(noof_customers, service_time_hrs, area_of_service_kms):
    service_time = service_time_hrs * 60 # minutes
    area_of_service = sqKm_to_sqMiles(area_of_service_kms) # square km
    filepath = f"DARPDATASET/Temportal-DS/nCustomers_{noof_customers}/Temporal_SD{service_time_hrs}hrs_SA{area_of_service_kms}km.txt"
    requests = []
    with open(filepath, 'r') as f:
        # first line is the depot
        for id, line in enumerate(f.readlines()):
            requests.append(Request(id, line, service_time, int(area_of_service)))
    return requests

File: src/julia-v1/parseRequests.jl
const DROPOFF_ALPHA = 1.5
# Constants for reading data line
const IDX = 1
const SRC_POINT_X = 2
const SRC_POINT_Y = 3
const DST_POINT_X = 4
const DST_POINT_Y = 5
const DRT = 6
const MRT= 7
const PICKUP_OR_DROPOFF_TIME = 8
const ISPICK_TIME = 9

struct Point
    x::Int32
    y::Int32
    Point(x, y) = new(x, y)
end

struct Request
    load::Int32 # number of passengers for this request
    id::Int32
    src::Point
    dst::Point
    direct_ride_time::Float64
    max_ride_time::Float64
    pickup_time::Float64
    dropoff_time::Float64
    is_pickup::Bool
end

function request_from_dataline(line::AbstractString)
    parts = split(line, "\t")
    parts = filter(part -> part != "", parts)
    # the parts should ateleast of size 9
    if length(parts) < 9
        return nothing
    end

    time = parse(Float64, parts[PICKUP_OR_DROPOFF_TIME])
    drt = parse(Float64, parts[DRT])
    mrt = parse(Float64, parts[MRT])
    return Request(1,
                    parse(Int32, parts[IDX]),
                    Point(parse(Int32, parts[SRC_POINT_X]), parse(Int32, parts[SRC_POINT_Y])),
                    Point(parse(Int32, parts[DST_POINT_X]), parse(Int32, parts[DST_POINT_Y])),
                    drt,
                    mrt,
                    time,
                    time + (DROPOFF_ALPHA * drt),
                    Bool(parse(Int8, parts[ISPICK_TIME]))
                    )
end


function parseData(noofCustomers::Int64,
    serviceDuration::Int64, areaOfService::Int64)

    basepath = "/Users/raghuveernaraharisetti/mscs/dail-a-ride/Dail-A-Ride-TabuSearch-CPU"
    filepath = "$(basepath)/DARPDATASET/Temportal-DS/nCustomers_$(noofCustomers)/Temporal_SD$(serviceDuration)hrs_SA$(areaOfService)km.txt"
    print(filepath)
    requests = Request[]
    filelines = readlines(filepath)

    for line in filelines
        req = request_from_dataline(line)
        if isnothing(req)
            println("nothing")
            continue
        end
        push!(requests, req)
    end

    return requests
end

